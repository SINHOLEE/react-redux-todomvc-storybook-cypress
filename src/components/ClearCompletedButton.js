import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { clearCompletedTodos } from '../actions';

function ClearCompletedButton({ hasCompleted, clearCompletedTodos }) {
  if (!hasCompleted) {
    return null;
  }
  return (
    <button className="clear-completed" onClick={clearCompletedTodos}>
      Clear completed
    </button>
  );
}

ClearCompletedButton.propTypes = {
  hasCompleted: PropTypes.bool.isRequired,
  clearCompletedTodos: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  hasCompleted: state.todos.some(todo => todo.completed)
});

export default connect(
  mapStateToProps,
  { clearCompletedTodos }
)(ClearCompletedButton);
