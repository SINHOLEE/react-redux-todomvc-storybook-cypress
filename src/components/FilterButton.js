import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

function FilterButton({ type, isSelected }) {
  return (
    <Link
      disabled={isSelected}
      to={'/' + type}
      className={`filter-item ${isSelected ? 'selected' : ''}`}
    >
      {type}
    </Link>
  );
}

FilterButton.propTypes = {
  type: PropTypes.string.isRequired,
  isSelected: PropTypes.bool.isRequired
};

const mapStateToProps = (state, ownProps) => {
  return {
    isSelected: state.nowShowing === ownProps.type
  };
};

export default connect(mapStateToProps)(FilterButton);
