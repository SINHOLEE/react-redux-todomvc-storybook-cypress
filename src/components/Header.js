import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {keyCodes} from '../constants';
import {addTodo} from '../actions';

export class Header extends React.Component {
  static propTypes = {
    addTodo: PropTypes.func.isRequired
  };

  state = {text: ''};

  onKeyDown(ev) {
    if (ev.keyCode === keyCodes.ENTER) {
      const text = this.state.text.trim();

      if (text) {
        this.props.addTodo(text);
        this.setState({text: ''});
      }
    }
  }

  onChange(ev) {
    this.setState({
      text: ev.target.value
    });
  }

  render() {
    return (
      <header className="header">
        <h1>todos</h1>
        <input
          data-testid="todo-input"
          className="new-todo"
          placeholder="What needs to be done?"
          value={this.state.text}
          onKeyDown={ev => this.onKeyDown(ev)}
          onChange={ev => this.onChange(ev)}
        />
      </header>
    );
  }
}

export default connect(
  null,
  {addTodo}
)(Header);
