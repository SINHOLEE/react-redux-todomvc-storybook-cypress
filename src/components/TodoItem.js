import React from 'react';
import PropTypes from 'prop-types';
import { keyCodes } from '../constants';
import { connect } from 'react-redux';
import {
  addTodo,
  removeTodo,
  updateTodo,
  toggleTodo,
  setEditing
} from '../actions';

export class TodoItem extends React.Component {
  static propTypes = {
    todo: PropTypes.object.isRequired,
    toggleTodo: PropTypes.func.isRequired,
    setEditing: PropTypes.func.isRequired,
    removeTodo: PropTypes.func.isRequired,
    updateTodo: PropTypes.func.isRequired,
    isEditing: PropTypes.bool.isRequired
  };

  componentDidUpdate() {
    const { isEditing } = this.props;

    if (isEditing) {
      const textLength = this.input.value.length;

      this.input.focus();
      this.input.setSelectionRange(textLength, textLength);
    }
  }

  save() {
    const { updateTodo, todo, setEditing } = this.props;
    const text = this.input.value.trim();

    if (text) {
      updateTodo(todo.id, text);
    }
    setEditing(null);
  }

  onKeyDown(ev) {
    switch (ev.keyCode) {
      case keyCodes.ENTER:
        this.save();
        break;
      case keyCodes.ESCAPE:
        this.props.setEditing(null);
        break;
      default:
        return;
    }
  }

  render() {
    const { todo, toggleTodo, setEditing, removeTodo, isEditing } = this.props;
    const { id, text, completed } = todo;
    const liClassName = `${completed ? 'completed' : ''} ${
      isEditing ? 'editing' : ''
    }`;

    return (
      <li className={liClassName}>
        <div className="view">
          <input
            className="toggle"
            type="checkbox"
            checked={completed}
            onChange={() => toggleTodo(id)}
          />
          <label onDoubleClick={() => setEditing(id)}>{text}</label>
          <button className="destroy" onClick={() => removeTodo(id)} />
        </div>
        <input
          className="edit"
          defaultValue={text}
          onKeyDown={ev => this.onKeyDown(ev)}
          onBlur={() => this.save()}
          ref={input => (this.input = input)}
        />
      </li>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  isEditing: ownProps.todo.id === state.editing
});

const mapDispatchToProps = {
  addTodo,
  removeTodo,
  updateTodo,
  toggleTodo,
  setEditing
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TodoItem);
