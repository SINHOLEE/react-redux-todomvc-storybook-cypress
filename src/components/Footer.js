import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import FilterButton from './FilterButton';
import ClearCompletedButton from './ClearCompletedButton';
import { listFilters } from '../constants';

function Footer({ leftCount }) {
  return (
    <footer className="footer">
      <span className="todo-count">
        <strong>{leftCount}</strong>
        <span> items</span>
        <span> left</span>
      </span>
      <ul className="filters">
        <li>
          <FilterButton type={listFilters.ALL} />
        </li>
        <li>
          <FilterButton type={listFilters.ACTIVE} />
        </li>
        <li>
          <FilterButton type={listFilters.COMPLETED} />
        </li>
      </ul>
      <ClearCompletedButton />
    </footer>
  );
}

Footer.propTypes = {
  leftCount: PropTypes.number.isRequired
};

const mapStateToProps = state => ({
  leftCount: state.todos.filter(todo => !todo.completed).length
});

export default connect(mapStateToProps)(Footer);
