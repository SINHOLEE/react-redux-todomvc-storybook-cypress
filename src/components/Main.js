import React from 'react';
import PropTypes from 'prop-types';
import TodoItem from './TodoItem';
import { connect } from 'react-redux';
import { listFilters } from '../constants';
import { toggleAllTodos } from '../actions';

function Main({ todos, toggleAllTodos }) {
  return (
    <section className="main">
      <input type="checkbox" className="toggle-all" onClick={toggleAllTodos} />
      <ul className="todo-list">
        {todos.map(todo => (
          <TodoItem key={todo.id} todo={todo} />
        ))}
      </ul>
    </section>
  );
}

Main.propTypes = {
  todos: PropTypes.array.isRequired
};

function getFilteredTodos(todos, nowShowing) {
  switch (nowShowing) {
    case listFilters.ACTIVE:
      return todos.filter(todo => !todo.completed);
    case listFilters.COMPLETED:
      return todos.filter(todo => todo.completed);
    case listFilters.ALL:
    default:
      return todos;
  }
}

const mapStateToProps = (state, ownProps) => {
  return { todos: getFilteredTodos(state.todos, state.nowShowing) };
};

const mapDispatchToPorps = { toggleAllTodos };

export default connect(
  mapStateToProps,
  mapDispatchToPorps
)(Main);
