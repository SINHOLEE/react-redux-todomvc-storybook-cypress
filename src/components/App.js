import React from 'react';
import { connect } from 'react-redux';
import { setFilter } from '../actions';
import { listFilters } from '../constants';
import Main from './Main';
import Header from './Header';
import Footer from './Footer';
import './app.css';

function getFilter(props) {
  return props.match.params.nowShowing || listFilters.ALL;
}

class App extends React.Component {
  componentDidMount() {
    this.props.setFilter(getFilter(this.props));
  }

  componentDidUpdate() {
    this.props.setFilter(getFilter(this.props));
  }

  render() {
    return (
      <div className="todoapp">
        <Header />
        <Main />
        <Footer />
      </div>
    );
  }
}

export default connect(
  null,
  { setFilter }
)(App);
