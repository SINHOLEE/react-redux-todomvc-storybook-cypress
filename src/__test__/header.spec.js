import React from 'react';
import { render } from 'react-dom';
import prettyHTML from 'diffable-html';
import { Header } from '../components/header';
import renderer from 'react-test-renderer';

it('Header component', () => {
  const el = document.createElement('div');
  render(<Header addTodo={() => {}} />, el);

  const outputHTML = prettyHTML(`
    <header class="header">
      <h1>todos</h1>
      <input class="new-todo" placeholder="What needs to be done?" value="" />
    </header>
  `);

  expect(prettyHTML(el.innerHTML)).toBe(outputHTML);
});

it('Header component - Snapshot', () => {
  const el = document.createElement('div');
  render(<Header addTodo={() => {}} />, el);

  expect(prettyHTML(el.innerHTML)).toMatchSnapshot();
});

it('Header component - Snapshot', () => {
  const tree = renderer.create(<Header />).toJSON();

  expect(tree).toMatchSnapshot();
});
