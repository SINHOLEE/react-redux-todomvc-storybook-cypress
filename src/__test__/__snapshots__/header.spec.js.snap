// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component - Snapshot 1`] = `
"
<header class=\\"header\\">
  <h1>
    todos
  </h1>
  <input class=\\"new-todo\\"
         placeholder=\\"What needs to be done?\\"
         value
  >
</header>
"
`;

exports[`Header component - Snapshot 2`] = `
<header
  className="header"
>
  <h1>
    todos
  </h1>
  <input
    className="new-todo"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="What needs to be done?"
    value=""
  />
</header>
`;
